### DISCUSSION TITLE
Let's scrap waffle flags from whoami. Who's with me?

### DISCUSSION THREAD

**Discussion by peterbe** [mdn/yari#3753] [Ideas]
<img width="782" alt="Screen Shot 2021-05-07 at 3 14 08 PM" src="https://user-images.githubusercontent.com/26739/117497751-e0432380-af46-11eb-8ef2-50270a2436f0.png">

Right now we don't use any waffle flags for anything. ...outside of Kuma. 
We do use a waffle Switch for [`welcome_email`](https://developer.mozilla.org/admin/waffle/switch/19/change/) which determines whether we send a welcome email or not [here](https://github.com/mdn/kuma/blob/12bf2afa3154e8df5a8f8ac92d6e2cc806bd14e4/kuma/users/signal_handlers.py#L68) and [here](https://github.com/mdn/kuma/blob/12bf2afa3154e8df5a8f8ac92d6e2cc806bd14e4/kuma/users/signal_handlers.py#L104). But let's not let that distract us. 

There is, today, no active uses of any of the `waffle` struct we get from Kuma to Yari in the `/api/v1/whoami`. 
Suppose that we want to launch another Developer Needs-like survey in a near future, what we probably would do is to manually "hardcode" it into Yari. Because we re-do this a bit it could be as simple as commenting in or out some existing code. 
Suppose our partners want the survey to go online on Febjuni 32nd, we can simply time it with a PR on Yari and wait for the next Yari release. Or if they want it to go out at exactly 16:00GMT we can simply kick off a Yari prod build at 15:00GMT and it'll be perfectly time. (or even better, land it in advance but put a `if (new Date().getTime() > TIME_TO_SHOW) {` snippet into the code (assuming we can trust the client timestamp)).
Suppose we want to do a thing where we only want it for 25% of the signed in population>? Easy, you just write (`if (userData.isAuthenticated && Math.random() > 0.75) { localStorage('enrolled_experiment', 'true') }`. And when you're done with that you delete the code from Yari (could be as simple as reverting a PR). 

The only thing Kuma's waffle Flag can do that Yari can't is to store your enrollment centrally for signed-in people. 

The other reason for removing it from Kuma is because it's probably just there because it couldn't be done in any other way back in the day when we didn't move nearly as fast. These days we can have a new version of Yari out in a matter of 1-2h. And equally, if it's on the Kuma side where we just land a PR and Ryan rolls out a release which is nearly all automated now. 

Last but not least, we have our users hit `/api/v1/whoami` a **LOT**. 
233M times in the last month. The, by far, most expensive thing in the `/api/v1/whoami` is to wake up the Elasticache to make 2 queries to see if the waffle flags apply to each and every single anonymous request.
<img width="1085" alt="Screen Shot 2021-05-07 at 3 29 52 PM" src="https://user-images.githubusercontent.com/26739/117499335-141f4880-af49-11eb-9706-1979b0ca912a.png">
It does slow down the endpoint. That waffle flag "induced" part is the slowest part of the Django view for anonymous users. If we can remove that we can more quickly return on the XHR request that Yari does on every request. 

Who's with me? 
I propose we leave waffle flags entirely as they are in all Kuma code, except within the `whoami` Django view function (and its tests). Then we remove it from Yari [here](https://github.com/mdn/yari/blob/be5a341f056e5347aefc7dd2d01bb78ae3d8f82a/client/src/user-context.tsx#L95) and review any existing old (inactive) Banner components that might refer to it.

**Comment by peterbe** [comment]
@schalkneethling @Gregoor @escattone ?
We can turn this into 2 issues on Kuma and here in Yari.

**Comment by escattone** [comment]
I'm all in! 😄

**Comment by peterbe** [comment]
Wheels are in motion! https://github.com/mdn/kuma/pull/7883
Once that lands, we need to start another one in Yari that does away with this code: https://github.com/mdn/yari/blob/ab633908cc71080f31f1545f9c5917f143f95a4b/client/src/user-context.tsx#L20

**Comment by peterbe** [comment]
IT WORKED! https://github.com/mdn/kuma/pull/7883#issuecomment-859534317

I'm just going to copy the same screenshot here because it's so pretty:

<img width="1217" alt="Screen Shot 2021-06-11 at 8 03 52 AM" src="https://user-images.githubusercontent.com/26739/121698710-4407c180-ca9c-11eb-8094-325b58e225f9.png">

<img width="863" alt="Screen Shot 2021-06-11 at 8 03 24 AM" src="https://user-images.githubusercontent.com/26739/121698731-4833df00-ca9c-11eb-9ef7-06eb4568a727.png">
